### PR TITLE
Added : Anti-Pattern for reducing data before joins

### PR DESCRIPTION
**Pull request by PoulamiR1994**

### CODE REVIEW

**Comment by franklinWhaite** [comment]
Please consider de following case, the current implementation is not detecting the antipattern if it happens within a subquery 

```
SELECT
    tt.sum_col2, count(1) ct
FROM 
    (SELECT 
        t1.col1, sum(t2.col2) sum_col2
    FROM 
        t1
    JOIN
        t2 ON t1.col1 = t2.col2
    GROUP BY
        t1.col1 ) tt1
GROUP BY    
    tt1.sum_col2
```

File: bigquery-antipattern-recognition/src/main/java/com/google/zetasql/toolkit/antipattern/parser/visitors/IdentifyAggAfterJoinVisitor.java [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+import com.google.zetasql.toolkit.antipattern.util.ZetaSQLStringParsingHelper;
+import java.lang.reflect.Array;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.Set;
+
+public class IdentifyAggAfterJoinVisitor extends ParseTreeVisitor {
+
+  private final String SUGGESTION_MESSAGE =
+      "GROUP By found at line number %s after Join at line number %s.Try to apply aggregation before joining to reduce join data";
```
Comment (franklinWhaite) [inline bigquery-antipattern-recognition/src/main/java/com/google/zetasql/toolkit/antipattern/parser/visitors/IdentifyAggAfterJoinVisitor.java:39]: GROUP **BY** found at line number %s after **JOIN** at line number %s. **Consider applying** aggregation before joining to reduce join data";

File: bigquery-antipattern-recognition/src/main/java/com/google/zetasql/toolkit/antipattern/parser/visitors/IdentifyAggAfterJoinVisitor.java [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+
+      foundGroupBy = true;
+      groupByLocation = ZetaSQLStringParsingHelper.countLine(query, selectNode.getGroupBy().getParseLocationRange().start());
+      for (ASTGroupingItem groupingItem : selectNode.getGroupBy().getGroupingItems()) {
+        ASTNodes.ASTPathExpression groupPath = (ASTNodes.ASTPathExpression) groupingItem.getExpression();
+
+        if (groupPath.getNames().size() > 1) {
+          groupByKeys.add(groupPath.getNames().get(1).getIdString());
+        } else {
+          groupByKeys.add(groupPath.getNames().get(0).getIdString());
```
Comment (franklinWhaite) [inline bigquery-antipattern-recognition/src/main/java/com/google/zetasql/toolkit/antipattern/parser/visitors/IdentifyAggAfterJoinVisitor.java:91]: What does the first entry of groupPath.getNames() correspond to, why do we need this value?

File: bigquery-antipattern-recognition/src/main/java/com/google/zetasql/toolkit/antipattern/parser/visitors/IdentifyAggAfterJoinVisitor.java [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+  @Override
+  public void visit(ASTNodes.ASTBinaryExpression node) {
+
+    ASTNodes.ASTPathExpression lhsPath = (ASTNodes.ASTPathExpression) node.getLhs();
+    ASTNodes.ASTPathExpression rhsPath = (ASTNodes.ASTPathExpression) node.getRhs();
+
+    if (lhsPath.getNames().size() > 1) {
+      lhsJoinKeys.add(lhsPath.getNames().get(1).getIdString());
+    } else {
+      lhsJoinKeys.add(lhsPath.getNames().get(0).getIdString());
```
Comment (franklinWhaite) [inline bigquery-antipattern-recognition/src/main/java/com/google/zetasql/toolkit/antipattern/parser/visitors/IdentifyAggAfterJoinVisitor.java:110]: What does the first entry of lhsPath.getNames() correspond to, why do we need this value?

File: bigquery-antipattern-recognition/src/main/java/com/google/zetasql/toolkit/antipattern/parser/visitors/IdentifyAggAfterJoinVisitor.java (hunk 2 above)
Comment (PoulamiR1994) [reply bigquery-antipattern-recognition/src/main/java/com/google/zetasql/toolkit/antipattern/parser/visitors/IdentifyAggAfterJoinVisitor.java:91]: When GROUP BY uses aliases, the first entry corresponds to the alias, else it corresponds to the column name. This is to ensure we are getting only the column name.

File: bigquery-antipattern-recognition/src/main/java/com/google/zetasql/toolkit/antipattern/parser/visitors/IdentifyAggAfterJoinVisitor.java (hunk 3 above)
Comment (PoulamiR1994) [reply bigquery-antipattern-recognition/src/main/java/com/google/zetasql/toolkit/antipattern/parser/visitors/IdentifyAggAfterJoinVisitor.java:110]: When join columns use aliases, the first entry corresponds to the alias, else it corresponds to the column name. This is to ensure we are getting only the column name.

Comment (franklinWhaite) [reply bigquery-antipattern-recognition/src/main/java/com/google/zetasql/toolkit/antipattern/parser/visitors/IdentifyAggAfterJoinVisitor.java:120]: it seems to me that you want the last record of the list. If so consider a syntax like the following 

```
ImmutableList<ASTIdentifier> namesList = lhsPath.getNames();
lhsJoinKeys.add(namesList.get(namesList.size()-1).getIdString());
```

File: bigquery-antipattern-recognition/src/main/java/com/google/zetasql/toolkit/antipattern/parser/visitors/IdentifyAggAfterJoinVisitor.java (hunk 2 above)
Comment (franklinWhaite) [reply bigquery-antipattern-recognition/src/main/java/com/google/zetasql/toolkit/antipattern/parser/visitors/IdentifyAggAfterJoinVisitor.java:92]: it seems to me that you want the last record of the list. If so consider a syntax like the following 

```
ImmutableList<ASTIdentifier> namesList = lhsPath.getNames();
lhsJoinKeys.add(namesList.get(namesList.size()-1).getIdString());
```

File: bigquery-antipattern-recognition/src/main/java/com/google/zetasql/toolkit/antipattern/parser/visitors/IdentifyAggAfterJoinVisitor.java (hunk 3 above)
Comment (PoulamiR1994) [reply bigquery-antipattern-recognition/src/main/java/com/google/zetasql/toolkit/antipattern/parser/visitors/IdentifyAggAfterJoinVisitor.java]: Added and tested the change suggested above.

File: bigquery-antipattern-recognition/src/main/java/com/google/zetasql/toolkit/antipattern/parser/visitors/IdentifyAggAfterJoinVisitor.java (hunk 2 above)
Comment (PoulamiR1994) [reply bigquery-antipattern-recognition/src/main/java/com/google/zetasql/toolkit/antipattern/parser/visitors/IdentifyAggAfterJoinVisitor.java:93]: Added and tested the change suggested above.

**Pull request closed without merging**